rl_algorithms/distillation/dqn_agent.py

```diff
@@ -32,83 +35,162 @@ class DistillationDQN(DQNAgent):
     # pylint: disable=attribute-defined-outside-init
     def _initialize(self):
         """Initialize non-common things."""
-        self.softmax_tau = 0.01
-        self.learner = build_learner(self.learner_cfg)
 
-        self.buffer_path = (
-            f"./data/distillation_buffer/{self.log_cfg.env_name}/"
-            + f"{self.log_cfg.agent}/{self.log_cfg.curr_time}/"
+        # You must choose one of them.
+        assert (
+            self.args.teacher
+            + self.args.student
+            + self.args.add_expert_q
+            + self.args.test
+            == 1
         )
-        if self.args.distillation_buffer_path:
-            self.buffer_path = "./" + self.args.distillation_buffer_path
-        os.makedirs(self.buffer_path, exist_ok=True)
 
-        self.memory = DistillationBuffer(
-            self.hyper_params.batch_size, self.buffer_path, self.log_cfg.curr_time,
-        )
+        if self.args.student or self.args.test:
+            # Training student or generating distillation data(test) requires DistillationBuffer.
+
+            self.softmax_tau = 0.01
+            self.learner = build_learner(self.learner_cfg)
+            self.buffer_path = self.hyper_params.buffer_path
+            if not self.args.student:
+                self.buffer_path = (
+                    f"./data/distillation_buffer/{self.log_cfg.env_name}/"
+                    + f"{self.log_cfg.agent}/{self.log_cfg.curr_time}/"
+                )
+                if self.args.distillation_buffer_path:
+                    self.buffer_path = "./" + self.args.distillation_buffer_path
+                os.makedirs(self.buffer_path, exist_ok=True)
+
+            self.memory = DistillationBuffer(
+                self.hyper_params.batch_size, self.buffer_path, self.log_cfg.curr_time,
+            )
+        else:
+            # Since raining teacher do not require DistillationBuffer,
+            # it overloads DQNAgent._initialize.
+
+            DQNAgent._initialize(self)
+            self.save_distillation_dir = (
+                "./data/distillation_buffer/"
+                + self.env_info.name
+                + time.strftime("/%Y%m%d%H%M%S/")
+            )
+            os.makedirs(self.save_distillation_dir)
+            self.save_count = 0
 
-    def select_action(self, state: np.ndarray) -> np.ndarray:
+    def select_action(self, state: np.ndarray, is_test: bool = False) -> np.ndarray:
         """Select an action from the input space."""
-        self.curr_state = state
-        # epsilon greedy policy
-        # pylint: disable=comparison-with-callable
-        state = self._preprocess_state(state)
-        q_values = self.learner.dqn(state)
-
-        if not self.args.test and self.epsilon > np.random.random():
-            selected_action = np.array(self.env.action_space.sample())
+
+        if self.args.teacher:
```

Comment:
should do refactoring:
- move save logic to train method

Reply:
resolved in e745842